game/src/edit/traffic_signals/edits.rs

```diff
@@ -0,0 +1,211 @@
+                        .align_right(),
+                ]),
+                Widget::row(vec![
+                    "Seconds:".draw_text(ctx),
+                    Spinner::new(
+                        ctx,
+                        (5, 300),
+                        current.simple_duration().inner_seconds() as isize,
+                    )
+                    .named("duration"),
+                ]),
+                Widget::row(vec![
+                    "Type:".draw_text(ctx),
+                    Checkbox::toggle(
+                        ctx,
+                        "phase type",
+                        "fixed",
+                        "adaptive",
+                        None,
+                        match current {
+                            PhaseType::Fixed(_) => true,
+                            PhaseType::Adaptive(_) => false,
+                        },
+                    ),
+                ]),
+                Btn::text_bg2("Apply").build_def(ctx, hotkey(Key::Enter)),
+            ]))
+            .build(ctx),
+            idx,
+        })
+    }
+}
+
+impl State for ChangeDuration {
+    fn event(&mut self, ctx: &mut EventCtx, _: &mut App) -> Transition {
+        match self.composite.event(ctx) {
+            Outcome::Clicked(x) => match x.as_ref() {
+                "close" => Transition::Pop,
+                "Apply" => {
+                    let dt = Duration::seconds(self.composite.spinner("duration") as f64);
+                    let new_type = if self.composite.is_checked("phase type") {
+                        PhaseType::Fixed(dt)
+                    } else {
+                        PhaseType::Adaptive(dt)
+                    };
+                    let idx = self.idx;
+                    return Transition::PopWithData(Box::new(move |state, ctx, app| {
+                        let editor = state.downcast_mut::<TrafficSignalEditor>().unwrap();
+                        editor.add_new_edit(ctx, app, idx, |ts| {
+                            ts.phases[idx].phase_type = new_type.clone();
+                        });
+                    }));
+                }
+                _ => unreachable!(),
+            },
+            _ => {
+                if ctx.normal_left_click() && ctx.canvas.get_cursor_in_screen_space().is_none() {
+                    return Transition::Pop;
+                }
+                Transition::Keep
+            }
+        }
+    }
+
+    fn draw_baselayer(&self) -> DrawBaselayer {
+        DrawBaselayer::PreviousState
+    }
+
+    fn draw(&self, g: &mut GfxCtx, _: &App) {
+        self.composite.draw(g);
+    }
+}
+
+pub fn edit_entire_signal(
+    ctx: &mut EventCtx,
+    app: &App,
+    i: IntersectionID,
+    mode: GameplayMode,
+    original: BundleEdits,
+) -> Box<dyn State> {
+    let has_sidewalks = app
+        .primary
+        .map
+        .get_turns_in_intersection(i)
+        .any(|t| t.between_sidewalks());
+
+    let use_template = "use template";
```

Comment:
I hadn't seen these fancy editing options before. :+1:

Reply:
Just a few days ago, I finally cleaned up `ChooseSomething`! The API used to be this very complicated sequence-of-prompts "wizard" that kind of mimicked coroutines.